Game menu

octopusinvitro opened the pull request:
### Summary

The switch statement was replaced by a hash map and a method for every menu option.
There is now an initialization method that sets the menu options and the printed messages in the same place.
From my ignorant point of view, it is still not 100% OCP friendly, because you can add a method for every new option, but you have to _modify_ the initialization method... although the modification is to add two new lines, so, not sure about this.
### Detailed summary

Get rid of the switch statement in the menu class:
- Get rid of the enum.
- Use a Hash Map instead.
- Add new methods to create the players according to option value.

Made some functional Java to implement menu logic.
- The key of the Hash Map is the user option.
- Created a Hash Map whose value is a function.

Clean the game tests:
- Set the relevant methods to private in the game class.
- Remove unnecessary tests.
- Move the tests to the future test class.

Initialize the menu options in its own method. Also:
- Rename playerFunctions to menuOptions.
- Fix formatting.

Add the printed menu options to the menu class.
- Create an array list to add the printed messages for every option.
- Add a method to print the menu that uses the user interface instead of calling a method of the user interface.
- Remove the menu string in the user interface and modify the method that printed it to pass it the method to print.

Commit:
Get rid of the switch statement in the menu class:
 - Get rid of the enum.
 - Use a Hash Map instead.
 - Add new methods to create the players according to option value.

Commit:
Made some functional Java to implement menu logic.
- The key of the Hash Map is the user option.
- Created a Hash Map whose value is a function.

Commit:
Clean the game tests:
- Set the relevant methods to private in the game class.
- Remove unnecessary tests.
- Move the tests to the future test class.

Commit:
Initialize the menu options in its own method. Also:
- Rename playerFunctions to menuOptions.
- Fix formatting.

Commit:
Add the printed menu options to the menu class.

- Create an array list to add the printed messages for every option.
- Add a method to print the menu that uses the user interface instead of calling a method of the user interface.
- Remove the menu string in the user interface and modify the method that printed it to pass it the method to print.

Commit:
Modify travis to add the right Java version.

ecomba commented on src/main/java/Menu.java:
```diff
@@ -1,40 +1,60 @@
-        TWO_HUMANS, HUMAN_AND_ROBOT, TWO_ROBOTS, HUMAN_AND_ALIEN
-    }
+    private HashMap<String, Supplier<List<Player>>> menuOptions;
+    private ArrayList<String> printedMenuOptions;
```
You would want to use `Map` (`java.util.Map`) and `List` (`java.util.List`) for the declaration of these. It is a good practice to use the interfaces instead of the concrete classes when you define your variables.

octopusinvitro replied:
Good catch :+1:

ecomba commented on src/main/java/Menu.java:
```diff
@@ -1,40 +1,60 @@
+
+        menuOptions.put("4", this::humanAndAlien);
+        printedMenuOptions.add(" 4) Human vs. alien\n");
+    }
```
Would it make sense to use Lambda expressions for the options as opposed to methods?

I am not sure about the use of the `String` for the key (1,2,3,4). Are these strings used anywhere else in the code? If so, can we remove that duplication?

octopusinvitro replied:
They are used in `setOption()`, where I get the option from the user input, and in `isInvalidOption(String option)` where I check that the option is correct. Will try to find a better way to handle it :+1:

ecomba commented on src/main/java/Menu.java:
```diff
@@ -50,11 +70,19 @@ private String setOption() {
+            menu += printedMenuOption;
+        }
+
+        return gameUI.printMenuAndGetOption(menu);
```
I am suspicious of methods that contain the word **And**. This is possibly due to the method not really doing just one thing only...

ecomba commented on src/main/java/UserInterface.java:
```diff
@@ -47,8 +41,8 @@ public void printWelcomeMessage() {
-        print(PLAYERSMENU);
+    public String printMenuAndGetOption(String menu) {
+        print(menu);
         return console.read();
```
See comment above. Is there a way to do this differently? Can we split the functionality and make sure it will do just one thing?

octopusinvitro replied:
Can't think of a way of splitting it, since both lines of code are related. Maybe it's better to just rename it to "`getMenuOption()`"?

ecomba commented on src/test/java/TurnTest.java:
```diff
@@ -0,0 +1,83 @@
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertTrue;
+
+public class TurnTest {
```
Consider either removing this test or implementing the examples...

Commit:
Adds PR suggestions:
- USe the List and Map interfaces instead of the implementors.
- Rename the method that prompts the user for a menu option.
- Still thinking of the best way to provide the menu options.
